### PR TITLE
OM-9 Added economic unit handling

### PR DESCRIPTION
**Pull request by malinowskikam**
[https://openimis.atlassian.net/browse/OM-9](https://openimis.atlassian.net/browse/OM-9)

### CODE REVIEW

File: msystems/services.py [hunk 1]
```diff
@@ -1,26 +1,49 @@
...
-        
-            return user
+            try:
+                user = self._get_or_create_user(username, user_data)
+                self._update_user_legal_entities(user, user_data)
+                return user
+            except BaseException as e:
+                # Extra logging for the development, should be removed for any real data usage
+                # as it will put personal information in logs
+                logger.error("Successful SAML login handling failed, username=%s, user_data=%s", username,
```
Comment (dborowiecki) [inline msystems/services.py]: you can change this for logger.debug, then will be skipped in production environment

File: msystems/services.py [hunk 2]
```diff
@@ -56,5 +78,45 @@ def _update_user(self, user: User, user_data: dict) -> None:
...
-        # TODO Handling of legal entities
-        pass
+        legal_entities = self._parse_legal_entities(user_data.get('OrganizationAdministrator'))
+        policyholders = [self._get_or_create_policy_holder(user, line[1], line[0]) for line in legal_entities]
+
+        self._delete_old_user_policyholders(user, policyholders)
+        self._add_new_user_policyholders(user, policyholders)
+
+    def _parse_legal_entities(self, legal_entities) -> map:
+        return map(lambda s: s.rsplit(' ', 1), legal_entities)
```
Comment (dborowiecki) [inline msystems/services.py]: It'd be good to add a comment explaining format of administrative organization string

File: msystems/views.py [hunk 3]
```diff
@@ -90,12 +83,29 @@ def acs(request):
...
+
+        if 'RelayState' in request.POST and _validate_relay_state(request.POST['RelayState']):
+            return redirect(auth.redirect_to(request.POST['RelayState']))
+        else:
+            return redirect(MsystemsConfig.base_login_redirect)
+    else:
+        logger.error("Logout attempt failed: %s\n%s", str(
+            errors[-1]), auth.get_last_error_reason())
+        # TODO Add information about failed login attempt for the user
+        return redirect(MsystemsConfig.base_login_redirect)
```
Comment (dborowiecki) [inline msystems/views.py]: I'd suggest reversing the condition:
```suggestion
    if errors:
        logger.error("Logout attempt failed: %s\n%s", str(
            errors[-1]), auth.get_last_error_reason())
        # TODO Add information about failed login attempt for the user
        return redirect(MsystemsConfig.base_login_redirect)
    
          username = auth.get_nameid()
      user_data = auth.get_attributes()
    
  logger.debug("Logout attempt, username=%s, user_data=%s", username, user_data)
  if 'RelayState' in request.POST and _validate_relay_state(request.POST['RelayState']):
      return redirect(auth.redirect_to(request.POST['RelayState']))
  else:
      return redirect(MsystemsConfig.base_login_redirect)
```

**Review by dborowiecki — CHANGES_REQUESTED** [review]
Left some comments

File: msystems/views.py (hunk 3 above)
Comment (malinowskikam) [reply msystems/views.py]: I removed the sls code (To be implemented in different ticket, also different approach), but i applied this to the acs endpoint

File: msystems/services.py (hunk 2 above)
Comment (malinowskikam) [reply msystems/services.py]: I added an explanation

File: msystems/services.py (hunk 1 above)
Comment (malinowskikam) [reply msystems/services.py]: I changed it to debug

**Review by dborowiecki — APPROVED** [review]

**Pull request merged**